sdk/core/azure-core/src/main/java/com/azure/core/http/HttpRequest.java

```diff
@@ -278,4 +289,102 @@ public HttpRequest copy() {
         final HttpHeaders bufferedHeaders = new HttpHeaders(headers);
         return new HttpRequest(httpMethod, url, bufferedHeaders, body);
     }
+
+    /**
+     * TODO (kasobol-msft) add docs.
+     * @param bufferingMode The buffering mode.
+     * @return The request.
+     */
+    public HttpRequest copyWithRetryableBody(HttpRequestBodyBufferingMode bufferingMode) {
+        try {
+            final HttpHeaders bufferedHeaders = new HttpHeaders(headers);
+            BinaryData retryableBody  = getRetryableBody(bufferingMode);
+            return new HttpRequest(httpMethod, url, bufferedHeaders, retryableBody);
+        } finally {
+            requestBodyCopied = true;
+        }
+    }
+
+    /**
+     * TODO (kasobol-msft) add docs.
+     * @param bufferingMode The buffering mode.
+     * @return The request.
+     */
+    public Mono<HttpRequest> copyWithRetryableBodyAsync(HttpRequestBodyBufferingMode bufferingMode) {
+        // TODO (kasobol-msft) make toBytes async in binary data to complete async behavior in final PR.
+        return Mono.fromCallable(() -> copyWithRetryableBody(bufferingMode));
+    }
+
+    private BinaryData getRetryableBody(HttpRequestBodyBufferingMode bufferingMode) {
+        if (body == null) {
+            return null;
+        }
+
+        if (bufferingMode == HttpRequestBodyBufferingMode.NEVER) {
+            return body;
+        }
+
+        BinaryDataContent content = BinaryDataHelper.getContent(body);
+
+        if (content instanceof ByteArrayContent
+            || content instanceof SerializableContent
+            || content instanceof StringContent
+            || content instanceof FileContent) {
+            return body;
+        } else if (content instanceof InputStreamContent) {
+            InputStream inputStream = content.toStream();
+            Long contentLength = getContentLength();
+            if (contentLength != null && contentLength < Integer.MAX_VALUE) {
+                if (inputStream.markSupported() && bufferingMode == HttpRequestBodyBufferingMode.SHALLOW) {
+                    if (requestBodyCopied) {
+                        try {
+                            inputStream.reset();
+                        } catch (IOException e) {
+                            throw LOGGER.logExceptionAsError(
+                                new UncheckedIOException("Unable to reset request body", e));
+                        }
+                    }
+                    inputStream.mark(contentLength.intValue());
+                    return body;
+                } else {
+                    // If stream is not seekable buffer body and set it in both derived and original request.
+                    byte[] bufferedContent = content.toBytes();
```

Comment:
Given this is in azure-core may I suggest a better internal API for this. Thoughts on providing an API that converts the `InputStream` into a `List<byte[]>` so we aren't allocating and resizing large (>100MB) byte arrays. We can also make each byte array a consistent size so future reseting, seeking, etc is trivial to calculate.

Reply:
Yup. We can also leverage fact that usually request has content length so we can allocate target bytearray directly.
There are various optimizations to do here. But I'd rather pursue them separately from this work. I.e. ship API first and then refine perf.

Reply:
Btw. I left this https://github.com/kasobol-msft/azure-sdk-for-java/blob/b98722592edd680ff4e444d640de21fc0c8604ee/sdk/core/azure-core/src/main/java/com/azure/core/http/HttpRequest.java#L314 . I'll see what can be done in the final PR, otherwise I'll log work item.